storage/innobase/handler/ha_innodb.cc

```diff
@@ -17116,6 +17121,74 @@ int innobase_rollback_by_xid(handlerton* hton, XID* xid)
 	}
 }
 
+#ifndef EMBEDDED_LIBRARY
+/**
+  This function is used to rollback one X/Open XA distributed transaction
+  which is in the prepared state asynchronously.
+
+  It only set the transaction's status to ACTIVE and persist the status.
+  The transaction will be rolled back by background rollback thread.
+
+  @param xid X/Open XA transaction identification
+
+  @return 0 or error number
+*/
+int innobase_recover_rollback_by_xid(const XID *xid)
+{
+  DBUG_EXECUTE_IF("innobase_xa_fail", return XAER_RMFAIL;);
+
+  if (high_level_read_only) {
+    return(XAER_RMFAIL);
+  }
+
+  /*
+    trx_get_trx_by_xid() sets trx's xid to null. Thus only one call for any
+    given XID can find the transaction. Subsequent calls by other threads
+    would return nullptr. That is what guarantees that no other thread can be
+    modifying the state of the transaction at this point.
+  */
+  trx_t *trx= trx_get_trx_by_xid(xid);
+  if (trx == NULL) {
+    return(XAER_RMFAIL);
+  }
+
+  // ddl should not be rolled back through recovery
+  ut_ad(!trx->dict_operation);
+  ut_ad(trx->is_recovered);
+  ut_ad(trx->state == TRX_STATE_PREPARED);
+
+#ifdef WITH_WSREP
+  ut_ad(!wsrep_is_wsrep_xid(&trx->xid));
+#endif
+
+  if (trx->rsegs.m_redo.undo) {
+    ut_ad(trx->rsegs.m_redo.undo->rseg == trx->rsegs.m_redo.rseg);
+
+    mtr_t mtr;
+    mtr.start();
+    trx_undo_set_state_at_prepare(trx, trx->rsegs.m_redo.undo, true, &mtr);
+    mtr.commit();
+
+    ut_ad(mtr.commit_lsn() > 0);
+  }
+
+  /* Change the trx state to TRX_STATE_ACTIVE. Background rollback thread
+     will rollback this trx later. */
+  trx->state= TRX_STATE_ACTIVE;
+  return 0;
+}
+
+void innobase_tc_log_recovery_done()
+{
+  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO
+      && !srv_read_only_mode) {
```

Comment:
log_buffer_flush_to_disk(true) should be called  if srv_read_only_mode is false. otherwise it will cause an assertion failure.

Reply:
You are right, it should be more like this:
```c++
  if (srv_read_only_mode)
    return;
  /* Make durable any innobase_recover_rollback_by_xid(). */
  log_buffer_flush_to_disk(true);
  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO)
  {
```
Even with `innodb_force_recovery=3` or `innodb_force_recovery=4` we expect that some `XA PREPARE` transactions may be changed back to "active" state, to note that they must be rolled back. We may choose not to complete the rollback during this server run, but a subsequent server invocation with a smaller value of `innodb_force_recovery` could perform that.

Reply:
Let’s revise the first condition to
```c++
  if (srv_force_recovery >= SRV_FORCE_NO_UNDO_LOG_SCAN || srv_read_only_mode)
    return;
```
in order to avoid any surprises on `innodb_force_recovery=5` or `innodb_force_recovery=6`. In these cases, no `XA PREPARE` transactions can be found, so no changes will need to be made durable either.

Reply:
fixed